### PR TITLE
Extract calypso-refined

### PR DESCRIPTION
**Pull request by asakaev**

### CODE REVIEW

File: modules/refined-tests/src/test/scala/ru/m2/calypso/refined/RefinedKeysSuite.scala [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import eu.timepit.refined.scalacheck.string._
+import eu.timepit.refined.types.string.NonEmptyString
+import org.scalatest.funsuite.AnyFunSuiteLike
+import org.scalatest.prop.Configuration
+import org.typelevel.discipline.scalatest.FunSuiteDiscipline
+import ru.m2.calypso.refined.MissingInstances._
+import ru.m2.calypso.testing.KeyCodecTests
+
+class RefinedKeysSuite extends AnyFunSuiteLike with FunSuiteDiscipline with Configuration {
+  checkAll("KeyCodec[NonEmptyString]", KeyCodecTests[NonEmptyString].codec)
```
Comment (gaponenko-andrei) [inline modules/refined-tests/src/test/scala/ru/m2/calypso/refined/RefinedKeysSuite.scala:12]: I've noticed we have `KeyEncoder[Int]`, but pretty much no build-in support for refined versions of it - no instances in `ru.m2.calypso.refined.MissingInstances`, no tests. Was just curious: is it intentional?

I could see why that would be the case as there are quite a few types we would maybe want to cover such as `NonNeg` and `PosInt`, which are perhaps not the most commonly used. Should we even have `KeyEncoder[Int]` in the first place then? I'm not sure.

Just something to think about if you haven't already.

File: README.md [hunk 2]
```diff
@@ -48,6 +48,18 @@ cats.data.NonEmptyList
...
+To use calypso-refined in an existing sbt project, add the following dependencies to your build.sbt:
+```scala
+libraryDependencies += "ru.m2" %% "calypso-refined" % "<version>"
+```
+```scala
+import eu.timepit.refined.types.string.NonEmptyString
+import ru.m2.calypso.refined._
+import ru.m2.calypso.syntax._
+
+NonEmptyString("Text").asBson // BsonString{value='Text'}
```
Comment (gaponenko-andrei) [inline README.md:60]: I might be missing something, but I'm not sure `NonEmptyString` has an apply signature to allow for this particular sintax 🤔 It doesn't seem to compile. Maybe something like the following would work slightly better instead?
```
    import eu.timepit.refined.collection.NonEmpty // or maybe import eu.timepit.refined.predicates.all._
    import eu.timepit.refined._
    import ru.m2.calypso.syntax._
    import ru.m2.calypso.refined._

    refineV[NonEmpty]("Text").asBson // BsonString{value='Text'}
```
Just a minor nitpick.

**Review by gaponenko-andrei — APPROVED** [review]

File: modules/refined/src/main/scala/ru/m2/calypso/refined/RefinedDecoder.scala [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import eu.timepit.refined.refineV
+import eu.timepit.refined.string.Uuid
+import ru.m2.calypso.Decoder
+
+trait RefinedDecoder {
+
+  implicit def decodeRefined[A: Decoder, P](implicit v: Validate[A, P]): Decoder[A Refined P] =
+    Decoder[A].emap(refineV(_))
+
+  implicit val decodeStringRefinedUuid: Decoder[String Refined Uuid] =
```
[the remark this answers is not in the document]
Comment (asakaev) [reply modules/refined/src/main/scala/ru/m2/calypso/refined/RefinedDecoder.scala:15]: Made an implementation with emap.

File: modules/refined-tests/src/test/scala/ru/m2/calypso/refined/RefinedKeysSuite.scala (hunk 1 above)
Comment (asakaev) [reply modules/refined-tests/src/test/scala/ru/m2/calypso/refined/RefinedKeysSuite.scala:12]: Yes, there was no need yet. When someone asks, we will add it.

**Pull request merged**